packager/packager.cc

```diff
@@ -0,0 +1,665 @@
+    auto rootKlass = ast::cast_tree<ast::ClassDef>(file.tree);
+    ENFORCE(rootKlass != nullptr);
+    auto moduleWrapper =
+        ast::MK::Module(core::LocOffsets::none(), core::Loc(ctx.file, core::LocOffsets::none()),
+                        name2Expr(packageMangledName, name2Expr(core::Names::Constants::PackageRegistry())), {},
+                        std::move(rootKlass->rhs));
+    rootKlass->rhs.clear();
+    rootKlass->rhs.emplace_back(move(moduleWrapper));
+    return file;
+}
+
+// We can't run packages without having all package ASTs. Assert that they are all present.
+bool checkContainsAllPackages(const core::GlobalState &gs, const vector<ast::ParsedFile> &files) {
+    UnorderedSet<core::FileRef> allPackages;
+    for (u4 i = 1; i < gs.filesUsed(); i++) {
+        core::FileRef fref(i);
+        if (fref.data(gs).sourceType == core::File::Type::Package) {
+            allPackages.insert(fref);
+        }
+    }
+
+    for (const auto &f : files) {
+        if (f.file.data(gs).sourceType == core::File::Type::Package) {
+            allPackages.erase(f.file);
+        }
+    }
+    return allPackages.empty();
+}
+
+} // namespace
+
+vector<ast::ParsedFile> Packager::run(core::GlobalState &gs, WorkerPool &workers, vector<ast::ParsedFile> files) {
+    Timer timeit(gs.tracer(), "packager");
+    // Ensure files are in canonical order.
+    fast_sort(files, [](const auto &a, const auto &b) -> bool { return a.file < b.file; });
+
+    // Step 1: Find packages and determine their imports/exports.
+    UnorderedMap<core::FileRef, shared_ptr<PackageInfo>> packageInfoByFile;
+    UnorderedMap<core::NameRef, shared_ptr<PackageInfo>> packageInfoByMangledName;
+    vector<shared_ptr<PackageInfo>> packages;
+    {
+        Timer timeit(gs.tracer(), "packager.findPackages");
+        core::UnfreezeNameTable unfreeze(gs);
+        for (auto &file : files) {
+            if (FileOps::getFileName(file.file.data(gs).path()) == PACKAGE_FILE_NAME) {
+                file.file.data(gs).sourceType = core::File::Type::Package;
+                core::MutableContext ctx(gs, core::Symbols::root(), file.file);
+                auto pkg = getPackageInfo(ctx, file);
+                // getPackageInfo emits an error if the package is malformed.
+                if (pkg != nullptr) {
+                    auto &existing = packageInfoByMangledName[pkg->name.mangledName];
+                    if (existing != nullptr) {
+                        if (auto e =
+                                ctx.beginError(pkg->loc.offsets(), core::errors::Packager::RedefinitionOfPackage)) {
+                            auto pkgName = absl::StrJoin(pkg->name.fullName, "::", NameFormatter(ctx));
+                            e.setHeader("Redefinition of package `{}`", pkgName);
+                            e.addErrorLine(existing->loc, "Package `{}` originally defined here", pkgName);
+                        }
+                    } else {
+                        existing = pkg;
```

Comment:
This is non-blocking, but I don't like indirect writes into elements of an unordered_map like this. It took me a long time to realize that this was allowed in C++ when I was first onboarding (e.g., I'd grep for `packageInfoByMangledName` and look for any lines that looked like they were writing into that element. Currently, the line initializing `existing` looks like it's only reading that element, so it's really tricky to use grep to figure out where anyone is ever writing into that map.

Consider an alternative:

```cpp
auto maybeExisting = packageInfoByMangledName.find(pkg->name.mangledName);
if (maybeExisting == packageInfoByMangledName.end()) {
  packageInfoByMangledName[pkg->name.mangledName] = pkg;
} else {
  auto existing = *maybeExisting;
  // ...
}
```

Reply:
I did it to save a lookup, but this isn't going to be a hot path ~ever, so I took your suggestion.